Add rule to ban usage of deprecated FormattedCompMessage component

wadimw opened the pull request.

wadimw committed:
Create subdirectory for rule tests

wadimw committed:
Create rule stub

wadimw committed:
Add detailed rule documetation

wadimw committed:
Add basic tests for Flow+JSX files

wadimw committed:
Implement rule logic

wadimw committed:
Add basic tests for files without Flow or JSX

wadimw committed:
Clean up test utils

wadimw committed:
Sort results by position in source file

wadimw committed:
Test content of actual result objects

wadimw committed:
Highlight only JSX element name

wadimw committed:
Enable rule in plugin and ruleset

wadimw committed:
Fix amount of expected rule exports in unit tests

ehoogerbeets commented on src/rules/BanFormattedCompMessage.js:
```diff
@@ -0,0 +1,125 @@
+    /** @type {Range} */ one,
+    /** @type {Range} */ other
+) => {
+    return one[0] === other[0] ? one[0] - other[0] : one[1] - other[1];
```
If the first elements are equal, then return the difference between them -- which will always be zero. Maybe this function isn't doing the right thing? I think you meant:

```
return one[0] === other[0] ? one[1] - other[1] : one[0] - other[0];
```

wadimw replied:
Right! I fixed it and refactored to be more verbose

ehoogerbeets commented on test/rules/BanFormattedCompMessage.test.js:
```diff
@@ -0,0 +1,431 @@
+                        "Do not use deprecated FormattedCompMessage component.",
+                    pathName: "x/y.js",
+                    rule,
+                    highlight: "Range: <e0>65:85</e0>"
```
We definitely need to merge my PR to get the highlight correctly!

ehoogerbeets commented on test/rules/BanFormattedCompMessage.test.js:
```diff
@@ -0,0 +1,431 @@
+            /** @type {string} */ filePath,
+            /** @type {string} */ content
+        ) => {
+            const parser = new JSParser();
```
The describe() says "Flow", but the parser being used is the regular JSParser.

wadimw replied:
Fixed

ehoogerbeets commented on test/utils.js:
```diff
@@ -0,0 +1,69 @@
+ * Remove a common indent from each line of a multiline string. Exact indent
+ * sequence is based on the indent of first non-empty line. Lines which don't
+ * begin with the detected indent are not modified.
+ */
```
Add in jsdocs for params and return value

wadimw replied:
Fixed

ehoogerbeets commented on test/utils.js:
```diff
@@ -0,0 +1,69 @@
+    items,
+    equalityComparer = (one, other) => one === other
+) => {
+    const /** @type {T[]} */ picked = [];
```
This also is O(n log n). Do we know if the items are already sorted when they are returned from the jsonpath query? If so, we can just check adjacent items to get rid of sequential dups which would be O(n).

wadimw replied:
Yeah this is a naive implementation because I figured performance is not going to be a concern here.

If you'd like, I can reimplement with a hashset which should have better performance (I think in JS Set.add() is optimistically O(1), though JS is not exactly a performance beast anyway).

wadimw replied:
Distinct filtering updated in 0a092de

wadimw replied:
Also moved it directly to the rule file, because I must have added it to test utils by accident

wadimw committed:
Fix range comparator

wadimw committed:
Fix incorrect parser in test suite

wadimw committed:
Update JSDoc for test util function

wadimw committed:
Improve distinct ranges filtering performance

Per MDN, Set.has() on average performs better than Array.includes().
Additionally, moved this function to rule file because it is not a test util.
